### PR TITLE
Reworks microfusion guns to behave like energy guns that can be upgraded, not like ballistic guns that happen to fire lasers

### PR DESCRIPTION
**Pull request by Paxilmaniac**
## About The Pull Request
MCRs happen to be on our list of very funny sec equipment that can become extremely strong as the round goes on, which would be fine if this wasn't paired with the fact that you can reload them the same way you could just a regular old ballistic gun. I love the idea of officers being able to customize their primary weapon to their liking, and I think we can keep that while also keeping them behaving like the energy guns they replaced with the following list of changes:

Cell related changes:
 - MCR cells spawn empty, and cannot be charged through cell chargers, or weapons charger.
 - MCR cells can be recharged when inserted into an MCR using a standard weapons charger.
 - MCR cells cannot be inserted into an MCR if they have charge, functionality to discharge cells outside of guns has been added to pair with this.
 - MCR cells will discharge all of their stored energy in a shower of sparks upon being removed from an MCR.
 - MCR cells put into something that isn't an MCR have a 10% chance to begin malfunctioning whenever charge is drawn from them.
 
Attachment related changes:
- Both of the scattershot upgrades will spread their damage out across 20 / number of lasers, instead of 30 and 45 respective to the two upgrades.
- The attachment that lets you change the color of your MCR bolts (should) be unobtainable through normal means
- The self charging attachment for cells has been moved to purchasing from micron control systems only, and cannot be printed.

Changes to Micron Control Systems:
 - The company's price has been changed to be on par with nanotrasen armories, instead of being one of the most expensive companies for very, very little benefit.
 - The price of ordering a fresh, unupgraded MCR from micron control systems through guncargo has been reduced to be equivalent to buying an allstar SC-2, that is to say, significantly cheaper.

Other stuff:
 - The cell malfunction warning will now tell you the correct time, 15 or whatever seconds instead of 150

## How This Contributes To The Skyrat Roleplay Experience
The goal of this pr is as the title says, to bring MCRs down to being upgradeable energy guns, rather than ballistic behaving guns that just so happen to fire lasers.
Going down the list for each of the changes:
Hotswapping cells is a cool idea until you realize that a sec officer can carry 100+ laser shots in a single cardboard box in their backpack, especially paired with how cheap MCR cells can be to print. Rather than remove the feature outright, researching better MCR cells allows the maximum charge you can carry to increase, while the mechanics for being unable to insert charged cells prevents the before mentioned 100 laser shots in a box situation.
Scattershot turned MCRs from decent guns into really good melee weapons with the fact that you get more overall damage if you shoot at someone in short range, which led to my beloathed strategy of dual wielding scattershot MCRs and running in a straight line at your target. Scattershot is now less of a damage booster, and more for simply letting players trade damage per shot for a higher chance to hit with multiple shots.
The beam color changer was cool until I saw that several officers, not naming names, would color their beams disabler blue despite the fact they are still fully lethal. Lol, lmao even, this is why we can't have nice things.
The self charging attachment, combined with the ability to hotswap cells, made it so officers could juggle multiple self charging cells for indefinite combat without having to worry about ever stopping to get more charge. This was mainly a problem due to the fact that they could just be printed and needed no extra work outside of that, so rather than remove it completely, it has been moved to being orderable only from micron control systems in guncargo.
Speaking of micron control systems, the company has been made significantly cheaper, it went practically unused because it could end up being somewhere around 24500 credits to unlock, roundstart. Even once you unlocked it, the MCRs themselves were extremely pricey, and that was without ANY modifications or upgrades. They are still expensive enough that it'd likely be better to just wait for RND to research the parts, but in a pinch it could help.

## Changelog

:cl:
balance: MCR cells are now upgrades rather than hotswappable 'laser magazines' that must be empty of charge before insertion into a gun
qol: MCR cells have a function to discharge themselves in case you end up with a charged one needing to be inserted into a gun
balance: MCR scattershot lasers will now do 20 damage divided by the number of lasers, rather than 30 or 45 divided
balance: The MCR laser coloring attachment is no longer attainable through lathes
balance: The MCR cell self charging attachment is only attainable through micron control systems in cargo
balance: Micron control systems is now equivalent in max price to NT armories, and the price of the MCR itself has been lowered to be equivalent to an allstar SC-2 (a little over half the original price)
fix: MCR cells will now tell you the correct number of seconds until they malfunction
/:cl:

### CODE REVIEW

**Comment by Stalkeros2** [comment]
I am fairly certain that the MCRs' readily hotswappable-precharged-essentially-magazines cells were one of their main, if not the main part of their design, and this PR violates that.

~~Not like I care that much to be mad or anything since bullets are subjectively cooler.~~

**Comment by Paxilmaniac** [comment]
> I am fairly certain that the MCRs' readily hotswappable-precharged-essentially-magazines cells were one of their main, if not the main part of their design, and this PR violates that.
> 
> ~Not like I care that much to be mad or anything since bullets are subjectively cooler.~

That was the original design, yes, but the original design went against everything an energy gun is designed to be, and because we went against what eguns are supposed to do, we ended up with an obscenely strong weapon that only gets stronger as research and the round progresses.

**Comment by Plum-but-gayer** [comment]
extremely based PR holy shit thank you

**Comment by KaiserBlackwood** [comment]
Flexing on the MCR users, Paxil? Good stuff.

**Comment by PCPort** [comment]
Finally. Balance has returned to the Force.

**Comment by Alanii** [comment]
> Cell related changes:
> 
> * MCR cells spawn empty, and cannot be charged through cell chargers, or weapons charger.
> * MCR cells can be recharged when inserted into an MCR using a standard weapons charger.
> * MCR cells cannot be inserted into an MCR if they have charge, functionality to discharge cells outside of guns has been added to pair with this.
> * MCR cells will discharge all of their stored energy in a shower of sparks upon being removed from an MCR.
> * MCR cells put into something that isn't an MCR have a 10% chance to begin malfunctioning whenever charge is drawn from them.

If you are going to do this you should, instead of being weird, change the firearm to have cells preinstalled and it can not be removed and remove the availability to print new cells. Otherwise make no sense and looks like a shit coded way to turn the MCR into every other energy gun. It would be smarter to actually change how the weapon functions instead of doing the "change" this way.

I foresee a lot of problems with how the MCRs will work in game due to the changes outlined, including how people will become confused the MCR does not work how its described to work.

tl;dr
If you are going to change the functionality of this firearm. Do it the correct way.

**Comment by Paxilmaniac** [comment]
> > Cell related changes:
> > 
> > * MCR cells spawn empty, and cannot be charged through cell chargers, or weapons charger.
> > * MCR cells can be recharged when inserted into an MCR using a standard weapons charger.
> > * MCR cells cannot be inserted into an MCR if they have charge, functionality to discharge cells outside of guns has been added to pair with this.
> > * MCR cells will discharge all of their stored energy in a shower of sparks upon being removed from an MCR.
> > * MCR cells put into something that isn't an MCR have a 10% chance to begin malfunctioning whenever charge is drawn from them.
> 
> If you are going to do this you should, instead of being weird, change the firearm to have cells preinstalled and it can not be removed and remove the availability to print new cells. Otherwise make no sense and looks like a shit coded way to turn the MCR into every other energy gun. It would be smarter to actually change how the weapon functions instead of doing the "change" this way.
> 
> I foresee a lot of problems with how the MCRs will work in game due to the changes outlined, including how people will become confused the MCR does not work how its described to work.
> 
> tl;dr If you are going to change the functionality of this firearm. Do it the correct way.

It is done this way because I still want people to have the ability to upgrade their mcrs, especially with cells and such. If players cannot gather from the fact that cells literally tell you in game that they discharge, cannot be inserted with charge, and so on, then I feel nothing in my power would help them.

**Comment by Paxilmaniac** [comment]
To second that, what I gather as your idea of 'doing it the correct way' is rewriting the whole of mcrs to do the exact same thing they already do, which is completely pointless.

**Comment by Ashrollen** [comment]
> Finally. Balance has returned to the Force.

Balance? There never will be a balance don't kid or joke on yourself.

File: modular_skyrat/modules/microfusion/code/microfusion_energy_master.dm [hunk 1]
```diff
@@ -619,7 +620,10 @@
 		if(reload_time && !HAS_TRAIT(user, TRAIT_INSTANT_RELOAD)) //This only happens when you're attempting a tactical reload, e.g. there's a mag already inserted.
 			if(display_message)
 				to_chat(user, span_notice("You start to insert [inserting_cell] into [src]!"))
-			if(!do_after(user, reload_time, src, IGNORE_USER_LOC_CHANGE))
+			if(inserting_cell.charge)
```
Comment (tf-4) [inline modular_skyrat/modules/microfusion/code/microfusion_energy_master.dm:621]: put this above the display message bit

File: modular_skyrat/modules/microfusion/code/microfusion_energy_master.dm [hunk 2]
```diff
@@ -630,7 +634,10 @@
 	else
 		if(display_message)
 			to_chat(user, span_notice("You start to insert [inserting_cell] into [src]!"))
-		if(!do_after(user, reload_time_slow, src, IGNORE_USER_LOC_CHANGE))
+		if(inserting_cell.charge)
```
Comment (tf-4) [inline modular_skyrat/modules/microfusion/code/microfusion_energy_master.dm:635]: again

**Review by tf-4 — CHANGES_REQUESTED** [review]
I'm struggling to work out why you would ever use more than one cell? Just remove the cell part of MCRs and make the previous cell attachments attach onto the gun itself.
This makes stuff like tac-reload and reloading attachment useless as well. Just remove those.

**Comment by Paxilmaniac** [comment]
> I'm struggling to work out why you would ever use more than one cell? Just remove the cell part of MCRs and make the previous cell attachments attach onto the gun itself. This makes stuff like tac-reload and reloading attachment useless as well. Just remove those.

You can still upgrade the cells using the better ones unlocked through tech, which will increase the max capacity of the gun of course

**Comment by tf-4** [comment]
eh ok

**Review by tf-4 — CHANGES_REQUESTED** [review]
remember to remove the ammo pouches and additional cells they come with, as they're now redundant

**Comment by Paxilmaniac** [comment]
> remember to remove the ammo pouches and additional cells they come with, as they're now redundant

do you know all the places that pouches and additional cells are given together?

**Comment by Stalkeros2** [comment]
https://www.youtube.com/watch?v=UcZzlPGnKdU

**Comment by Aether217** [comment]
> The attachment that lets you change the color of your MCR bolts (should) be unobtainable through normal means

This seems like an odd change that just removes customisation without affecting balance.

**Comment by GoldenAlpharex** [comment]
> > The attachment that lets you change the color of your MCR bolts (should) be unobtainable through normal means
> 
> This seems like an odd change that just removes customisation without affecting balance.

They mentioned how people were sadly using it to make their lasers be blue to look like non-lethals, when they _were_ actually lethals. It thus has a balance impact, because you're going to react differently to a blue laser than to a red one, due to blue being associated to non-lethal for everything but pulse rifles.

**Comment by Paxilmaniac** [comment]
It's an identifiability thing, you should be able to know at a glance that the projectile that is being fired at you is for certain gonna hurt.

File: modular_skyrat/modules/microfusion/code/microfusion_cell_attachments.dm [hunk 3]
```diff
@@ -123,49 +104,9 @@ If the cell isn't stabilised by a stabiliser, it may emit a radiation pulse.
 	microfusion_cell.self_charging = FALSE
 
 /obj/item/microfusion_cell_attachment/selfcharging/process_attachment(obj/item/stock_parts/cell/microfusion/microfusion_cell, delta_time)
-	if(microfusion_cell.charge < microfusion_cell.maxcharge)
-		microfusion_cell.charge = clamp(microfusion_cell.charge + (self_charge_amount * delta_time), 0, microfusion_cell.maxcharge)
-		if(microfusion_cell.parent_gun)
+	if(microfusion_cell.parent_gun)
```
Comment (tf-4) [inline modular_skyrat/modules/microfusion/code/microfusion_cell_attachments.dm]:
```suggestion
	if(!microfusion_cell.parent_gun)
		return
```
and deindent the rest but github review suggestions are shit

**Comment by IgiariValkyr** [comment]
-Can still print self-charging attachment
-Why would a cell immediately discharge all of its charge when the battery is taken out? This happens in zero other things that uses a cell.

I honestly agree with an above comment, just make the MCR have a pre-installed unremovable cell like _any other energy gun_ and attach the cell upgrades directly to the gun. If you only need one cell and can only ever use one cell, then why make cells printable at all? Change the "cell tiers" into attachments of their own.

Though this does remove EMP Microfusion Cell exploding, which is kinda nice. Now we can fight EMP blobs in relative peace without fear of our weapons exploding.

**Comment by weesnaw-ops** [comment]
thank Fuck

**Comment by Aether217** [comment]
> > > The attachment that lets you change the color of your MCR bolts (should) be unobtainable through normal means
> > 
> > 
> > This seems like an odd change that just removes customisation without affecting balance.
> 
> They mentioned how people were sadly using it to make their lasers be blue to look like non-lethals, when they _were_ actually lethals. It thus has a balance impact, because you're going to react differently to a blue laser than to a red one, due to blue being associated to non-lethal for everything but pulse rifles.

I don't consider it that great of an issue because there is still the sound indicator when the gun fires which is very unique to the MCRs unless changed, the inhand sprite and in my experience the beam itself will always be off colour and shaped differently.

Plus, you don't want to be hit by a projectile regardless of the colour, you aren't really going to act any different in combat based on that, you'd act based on the weapon that the player is using.

**Review by tf-4 — CHANGES_REQUESTED** [review]
Just make the colour changing thing operate on a preset list and don't add light blue

**Comment by Paxilmaniac** [comment]
“Just prohibit blue” isn’t going to do a thing for the problem I removed it for. It’s not just that they can look like disablers, it’s not having to play the guessing game of is this red | blue | purple | yellow | green | so on colored projectile about to kill me. While you can argue that the answer to that is almost always yes, that shouldn’t be a thing we make players have to wait until after they get hit to figure out just because some officers want rgb lasers.

**Comment by tf-4** [comment]
> “Just prohibit blue” isn’t going to do a thing for the problem I removed it for. It’s not just that they can look like disablers, it’s not having to play the guessing game of is this red | blue | purple | yellow | green | so on colored projectile about to kill me. While you can argue that the answer to that is almost always yes, that shouldn’t be a thing we make players have to wait until after they get hit to figure out just because some officers want rgb lasers.

In what cases are the yellow / purple / green lasers *not* going to kill you? While I understand the concern I don't see what there is to think about particularly. They all hurt.
This is not like allowing disabler-blue in which case you DO have to think about it and I'm in full agreement that that should be removed. Dark blue I can *see* an argument against even if I don't agree, but green? Yellow? Purple?

**Comment by Paxilmaniac** [comment]
Identifiability again, every default laser that hurts you is going to be red, there are exceptions to this of course, but you want to know when the exceptions happen? Special cases. If you see a laser that is a color other than red, that's supposed to tell you that there's something special about that projectile. Green laser? Oh shit its an xray laser! Yellow? While there aren't yellow lasers outside of CTF, yellow MCR lasers and .50 tracer rounds look close enough that the mistake can be made. Purple? Are they firing a plasma cutter at me? There are just so many chances for mistaken identity around that I'd much rather just have MCR lasers always be red without modification like other lasers similar to them, rather than have to constantly curate a list of colors that won't be mistaken for other things.

**Comment by tf-4** [comment]
good enough reasoning for me

**Comment by Zonespace27** [comment]
I said I wouldn't touch sec/antag stuff again but that doesn't stop me from merging this

**Comment by IgiariValkyr** [comment]
Still no changes on the batteries being useless now? And merged regardless? Boooooo
